src/parse.rs

```diff
@@ -90,128 +95,172 @@ enum Instr {
-    locals: Vec<String>,
-    code:   Vec<Instr>
+pub struct Bytecode {
+    classes: HashMap<String, Vec<Instr>>,
+    symbols: HashMap<(String, String), Fn>,
 }
 
-impl Method {
-    fn new() -> Method {
-        Method{
-            locals: Vec::new(),
-            code: Vec::new()
+impl Bytecode {
+    fn new(ctx : CompilerContext) -> Bytecode {
+        Bytecode {
+            classes: ctx.classes,
+            symbols: ctx.symbols
         }
     }
+}
+
+struct CompilerContext<'pt> {
+    classes: HashMap<String, Vec<Instr>>,
+    symbols: HashMap<(String, String), Fn>,
+
+    // Fields for convenience when building up the Bytecode struct
+    grm:        &'pt YaccGrammar,
+    input:      &'pt str,
+    cur_cls:    String,
+    cur_fn:     String,
+}
 
-    fn push_instr(&mut self, instr: Instr) {
-        self.code.push(instr)
+impl<'pt> CompilerContext<'pt> {
+    fn new(grm: &'pt YaccGrammar, input: &'pt str) -> CompilerContext<'pt> {
+        CompilerContext {
+            classes: HashMap::new(),
+            symbols: HashMap::new(),
+            grm:     grm,
+            input:   input,
+            cur_cls: "global".to_string(),
+            cur_fn:  "global".to_string(),
+        }
     }
 
-    fn push_var(&mut self, var_name: String) {
-        self.locals.push(var_name)
+    fn register_class(&mut self, class: &Node<u16>) {
+        match *class {
+            Node::Term { lexeme } => {
+                let class_name = self.get_value(class);
+                self.cur_cls   = class_name.clone();
+                self.classes.insert(class_name, Vec::new());
+            }
+            _ => panic!("Can only register a class on a terminal node")
+        }
     }
 
-}
+    fn register_function(&mut self, func: &Node<u16>) -> String {
+        match *func {
+            Node::Term { lexeme } => {
+                let func_name = self.get_value(func);
+                self.cur_fn = func_name.clone();
+                self.symbols.insert((self.cur_cls.to_string(), func_name.to_string()), Fn::new());
+                return func_name
+            }
+            _ => panic!("Can only register a func on a terminal node")
+        }
+    }
 
+    // Adds the parameter name to the param vector of the current cls + func.
+    fn register_parameter(&mut self, param: &Node<u16>) {
+        let param_name = self.get_value(param);
+        let ref key = (self.cur_cls.to_string(), self.cur_fn.to_string());
+        self.symbols.get_mut(key).unwrap().push_param(param_name);
+    }
 
-#[derive(Debug)]
-pub struct CompilerContext {
-    classes: HashMap<String,Class>,
-}
+    fn gen_bc(&mut self , instr: Instr) {
+        self.classes.get_mut(&self.cur_cls).unwrap().push(instr);
+    }
 
-impl CompilerContext {
-    fn add(&mut self, name: String, class: Class) {
-        self.classes.insert(name, class);
+    // Returns the last class which was registered. Useful context for building
+    // up the AST
+    fn current_class(&self) -> &str {
```

Comment:
I'm not convinced by the idea of this method ;) Also, nothing seems to use it?

Reply:
These fields are used internally by the `CompilerContext` struct. I think at one point I had a reason to expose them but that reason no longer exists. The accessor methods have been removed. Fixed :)